Don't repaint the outline bounds for (Legacy)RenderSVGModelObjects when their layout changes

heycam opened the pull request:
#### afe58473f11377d6373185ca3b3330898cf5feeb
<pre>
Don&apos;t repaint the outline bounds for (Legacy)RenderSVGModelObjects when their layout changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=242426">https://bugs.webkit.org/show_bug.cgi?id=242426</a>
&lt;rdar://problem/96574197&gt;

Reviewed by Simon Fraser.

LayoutRepainter repaints two rectangles: the &quot;clipped overflow rect&quot; and
the &quot;outline bounds&quot;. The outline bounds contains the CSS outline and any
box shadows. For SVG objects inheriting from (Legacy)RenderSVGModelObjects,
the CSS outline is already included in the clipped overflow rect (due to
it being included in SVGRenderSupport::computeFloatVisibleRectInContainer)
and SVG elements do not support box shadows, so for them, repainting the
outline bounds is wasted work.

Add an argument to LayoutRepainter to skip repainting the outline
bounds.

* Source/WebCore/rendering/LayoutRepainter.cpp:
(WebCore::LayoutRepainter::LayoutRepainter):
(WebCore::LayoutRepainter::repaintAfterLayout):
* Source/WebCore/rendering/LayoutRepainter.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::layout):
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::layout):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::layout):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::layout):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::layout):

Canonical link: <a href="https://commits.webkit.org/252244@main">https://commits.webkit.org/252244@main</a>
</pre>

smfr commented on Source/WebCore/rendering/LayoutRepainter.h:
```diff
@@ -32,9 +32,11 @@ namespace WebCore {
 class RenderElement;
 class RenderLayerModelObject;
 
+enum class RepaintOutlineBox : bool { No, Yes };
```
Can we call this `RepaintOutlineBounds` for consistency with `outlineBoundsForRepaint()`?

smfr commented on Source/WebCore/rendering/LayoutRepainter.h:
```diff
@@ -48,6 +50,7 @@ class LayoutRepainter {
     LayoutRect m_oldBounds;
     LayoutRect m_oldOutlineBox;
     bool m_checkForRepaint;
+    bool m_repaintOutlineBox;
```
Can we call this `m_repaintOutlineBounds ` for consistency with `outlineBoundsForRepaint()`?

heycam committed:
Don't repaint the outline bounds for (Legacy)RenderSVGModelObjects when their layout changes
https://bugs.webkit.org/show_bug.cgi?id=242426
<rdar://problem/96574197>

Reviewed by Simon Fraser.

LayoutRepainter repaints two rectangles: the "clipped overflow rect" and
the "outline bounds". The outline bounds contains the CSS outline and any
box shadows. For SVG objects inheriting from (Legacy)RenderSVGModelObjects,
the CSS outline is already included in the clipped overflow rect (due to
it being included in SVGRenderSupport::computeFloatVisibleRectInContainer)
and SVG elements do not support box shadows, so for them, repainting the
outline bounds is wasted work.

Add an argument to LayoutRepainter to skip repainting the outline
bounds.

* Source/WebCore/rendering/LayoutRepainter.cpp:
(WebCore::LayoutRepainter::LayoutRepainter):
(WebCore::LayoutRepainter::repaintAfterLayout):
* Source/WebCore/rendering/LayoutRepainter.h:
* Source/WebCore/rendering/svg/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::layout):
* Source/WebCore/rendering/svg/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::layout):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::layout):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::layout):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::layout):

Canonical link: https://commits.webkit.org/252244@main
